File: ltexls-core/src/main/java/org/bsplines/ltexls/parsing/latex/LatexCommandSignatureMatcher.java
/* Copyright (C) 2020 Julian Valentin, LTeX Development Community
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.bsplines.ltexls.parsing.latex;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.eclipse.xtext.xbase.lib.Pair;

public class LatexCommandSignatureMatcher {
  private List<LatexCommandSignature> commandSignatures;
  private Pattern commandPattern;
  private @Nullable String code;
  private @Nullable Matcher matcher;
  private @Nullable Set<String> ignoreCommandPrototypes;

  public LatexCommandSignatureMatcher(LatexCommandSignature commandSignature) {
    this(commandSignature, true);
  }

  public LatexCommandSignatureMatcher(LatexCommandSignature commandSignature,
        boolean escapeCommandName) {
    this(Collections.singletonList(commandSignature), escapeCommandName);
  }

  public LatexCommandSignatureMatcher(
        Collection<? extends LatexCommandSignature> commandSignatures) {
    this(commandSignatures, true);
  }

  public LatexCommandSignatureMatcher(
        Collection<? extends LatexCommandSignature> commandSignatures, boolean escapeCommandNames) {
    this.commandSignatures = new ArrayList<>(commandSignatures);

    Set<String> commandNames = new HashSet<>();
    this.commandSignatures.forEach((LatexCommandSignature x) -> commandNames.add(x.getName()));
    StringBuilder commandPatternStringBuilder = new StringBuilder("");
    boolean first = true;

    for (String commandName : commandNames) {
      if (first) {
        first = false;
      } else {
        commandPatternStringBuilder.append("|");
      }

      commandPatternStringBuilder.append(
          escapeCommandNames ? Pattern.quote(commandName) : commandName);
    }

    this.commandPattern = Pattern.compile(commandPatternStringBuilder.toString());

    this.code = null;
    this.matcher = null;
    this.ignoreCommandPrototypes = null;
  }

  public void startMatching(String code, Set<String> ignoreCommandPrototypes) {
    this.code = code;
    this.matcher = this.commandPattern.matcher(code);
    this.ignoreCommandPrototypes = new HashSet<String>(ignoreCommandPrototypes);
  }

  public @Nullable LatexCommandSignatureMatch findNextMatch() {
    // fixes false-positive dereference.of.nullable warnings
    @Nullable String code = this.code;
    @Nullable Matcher matcher = this.matcher;
    @Nullable Set<String> ignoreCommandPrototypes = this.ignoreCommandPrototypes;
    if ((code == null) || (matcher == null) || (ignoreCommandPrototypes == null)) return null;

    while (matcher.find()) {
      int fromPos = matcher.start();
      @Nullable LatexCommandSignatureMatch bestMatch = null;

      for (LatexCommandSignature commandSignature : this.commandSignatures) {
        if (ignoreCommandPrototypes.contains(commandSignature.getCommandPrototype())) {
          continue;
        }

        @Nullable List<Pair<Integer, Integer>> arguments =
            commandSignature.matchArgumentsFromPosition(code, fromPos);

        if (arguments != null) {
          LatexCommandSignatureMatch match = new LatexCommandSignatureMatch(
              commandSignature, code, fromPos, arguments);
          if ((bestMatch == null) || (match.getToPos() > bestMatch.getToPos())) bestMatch = match;
        }
      }

      if (bestMatch != null) return bestMatch;
    }

    return null;
  }

  public List<LatexCommandSignature> getCommandSignatures() {
    return Collections.unmodifiableList(this.commandSignatures);
  }
}

File: src/main/kotlin/org/bsplines/ltexls/parsing/latex/LatexCommandSignatureMatcher.kt
/* Copyright (C) 2019-2021 Julian Valentin, LTeX Development Community
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package org.bsplines.ltexls.parsing.latex

class LatexCommandSignatureMatcher(
  val commandSignatures: List<LatexCommandSignature>,
  escapeCommandPrefixes: Boolean = true,
) {
  private val commandRegex: Regex
  private var code: String? = null
  private var matchResultIterator: Iterator<MatchResult>? = null
  private var ignoreCommandPrototypes: Set<String>? = null

  init {
    val commandPatternStringBuilder = StringBuilder()
    var first = true

    for (commandSignature in commandSignatures) {
      if (first) {
        first = false
      } else {
        commandPatternStringBuilder.append("|")
      }

      val commandPrefix: String = commandSignature.prefix
      commandPatternStringBuilder.append(
        if (escapeCommandPrefixes) Regex.escape(commandPrefix) else commandPrefix
      )
    }

    commandRegex = Regex(commandPatternStringBuilder.toString())
  }

  fun startMatching(code: String, ignoreCommandPrototypes: Set<String>) {
    this.code = code
    this.matchResultIterator = this.commandRegex.findAll(code).iterator()
    this.ignoreCommandPrototypes = ignoreCommandPrototypes
  }

  fun findNextMatch(): LatexCommandSignatureMatch? {
    val code: String? = this.code
    val matchResultIterator: Iterator<MatchResult>? = this.matchResultIterator
    val ignoreCommandPrototypes: Set<String>? = this.ignoreCommandPrototypes

    if ((code == null) || (matchResultIterator == null) || (ignoreCommandPrototypes == null)) {
      return null
    }

    while (matchResultIterator.hasNext()) {
      val matchResult: MatchResult = matchResultIterator.next()
      val fromPos: Int = matchResult.range.first

      val lineStartPos: Int = code.lastIndexOf('\n', fromPos) + 1
      val precedingPartOfLine: String = code.substring(lineStartPos, fromPos)
      if (COMMENT_REGEX.find(precedingPartOfLine) != null) continue

      val bestMatch: LatexCommandSignatureMatch? =
          findBestMatchingCommandSignature(code, fromPos, ignoreCommandPrototypes)
      if (bestMatch != null) return bestMatch
    }

    return null
  }

  private fun findBestMatchingCommandSignature(
    code: String,
    fromPos: Int,
    ignoreCommandPrototypes: Set<String>,
  ): LatexCommandSignatureMatch? {
    var bestMatch: LatexCommandSignatureMatch? = null

    for (commandSignature in commandSignatures) {
      if (ignoreCommandPrototypes.contains(commandSignature.commandPrototype)) continue
      val arguments: List<Pair<Int, Int>>? =
        commandSignature.matchArgumentsFromPosition(code, fromPos)

      if (arguments != null) {
        val match = LatexCommandSignatureMatch(commandSignature, code, fromPos, arguments)
        if ((bestMatch == null) || (match.toPos > bestMatch.toPos)) bestMatch = match
      }
    }

    return bestMatch
  }

  companion object {
    private val COMMENT_REGEX = Regex("(?<!\\\\)(?:\\\\\\\\)*%")
  }
}
